lib/sycamore/sycamore/writers/pinecone_writer.py

```diff
@@ -0,0 +1,175 @@
+    def __init__(
+        self,
+        index_name: str,
+        index_spec: Union[None, dict, ServerlessSpec, PodSpec],
+        namespace: str,
+        dimensions: Optional[int],
+        distance_metric: str,
+        api_key: str,
+        batch_size: int = 100,
+    ):
+        self._index_name = index_name
+        self._index_spec = index_spec
+        self._namespace = namespace
+        self._dimensions = dimensions
+        self._distance_metric = distance_metric
+        self._batch_size = batch_size
+        self._api_key = api_key
+
+    def on_write_start(self) -> None:
+        from pinecone import Pinecone
+        from pinecone.core.client.exceptions import NotFoundException
+
+        pc = Pinecone(api_key=self._api_key)
+        try:
+            pc.describe_index(self._index_name)
+        except NotFoundException:
+            if self._dimensions is None:
+                raise ValueError(f"dimensions must be specified in order to create a new index {self._index_name}")
+            if self._index_spec is None:
+                raise ValueError(f"index spec must be supplied in order to create a new index {self._index_name}")
+            pc.create_index(
+                name=self._index_name,
+                dimension=self._dimensions,
+                spec=self._index_spec,
+                metric=self._distance_metric,
+                timeout=None,
+            )
+
+    def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+        from pinecone.grpc import PineconeGRPC
+
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
+
+        pc = PineconeGRPC(api_key=self._api_key)
+        index = pc.Index(self._index_name)
+
+        objects = self._extract_pinecone_objects(block)
+        obj_it = iter(objects)
+        batch = list(itertools.islice(obj_it, self._batch_size))
+        async_results = []
+        while len(batch) > 0:
+            async_results.append(index.upsert(vectors=batch, namespace=self._namespace, async_req=True))
+            batch = list(itertools.islice(obj_it, self._batch_size))
+        for r in async_results:
+            r.result()
+
+    @staticmethod
+    def _extract_pinecone_objects(block: Block) -> Iterable[Vector]:
+        def _flatten_metadata(data: Union[dict, list, tuple], prefix="") -> Iterable[Tuple[Any, Any]]:
+            iterator = []  # type: ignore
+            if isinstance(data, dict):
+                iterator = data.items()  # type: ignore
+            if isinstance(data, (list, tuple)):
+                iterator = enumerate(data)  # type: ignore
+            items = []
+            for k, v in iterator:
+                if isinstance(v, (dict, list, tuple)):
+                    if isinstance(v, (list, tuple)) and all(isinstance(innerv, str) for innerv in v):
+                        items.append(((str(k) if len(prefix) == 0 else f"{prefix}.{k}"), v))
```

Comment:
Minor, but might be a bit more readable if you defined a helper for `(str(k) if len(prefix) == 0 else f"{prefix}.{k}")` since it shows up a few times.